Tests/CacheAdvanceTests/CacheAdvanceTests.swift

```diff
@@ -96,6 +96,34 @@ final class CacheAdvanceTests: XCTestCase {
         XCTAssertEqual(messages, [])
     }
 
+  func test_messages_throwsFileCorruptedWhenOffsetInFileAtEndOfNewsetMessageOutOfSync() throws {
+      let header = try CacheHeaderHandle(
+        forReadingFrom: testFileLocation,
+        maximumBytes: 10_000,
+        overwritesOldMessages: true)
+      let cache = CacheAdvance<TestableMessage>(
+        fileURL: testFileLocation,
+        writer: try FileHandle(forWritingTo: testFileLocation),
+        reader: try CacheReader(forReadingFrom: testFileLocation),
+        header: try CacheHeaderHandle(
+            forReadingFrom: testFileLocation,
+            maximumBytes: header.maximumBytes,
+            overwritesOldMessages: header.overwritesOldMessages),
+        decoder: JSONDecoder(),
+        encoder: JSONEncoder())
+
+      // Make sure the header data is persisted before we read it as part of the `messages()` call below.
+      try header.synchronizeHeaderData()
+      // Our file is empty. Make the file corrupted by setting the offset at end of newest message to be further in the file.
+      // This should never happen, but past versions of this repo could lead to a file having this kind of inconsistency if a crash occurred at the wrong time.
+      try header.updateOffsetInFileAtEndOfNewestMessage(
+        to: FileHeader.expectedEndOfHeaderInFile + UInt64(MessageSpan.storageLength) + 1)
+
+      XCTAssertThrowsError(try cache.messages()) {
+          XCTAssertEqual($0 as? CacheAdvanceError, CacheAdvanceError.fileCorrupted)
```

Comment:
We should be able to detect this case and throw an appropriate error. We should also be able to prevent this case from happening in production. We solve both of these "shoulds" in the next few commits.